bgpd/bgp_zebra.c

```diff
@@ -2102,3 +2404,178 @@ int bgp_zebra_num_connects(void)
+void bgp_send_pbr_ipset_match(struct bgp_pbr_match *pbrim, bool install)
+{
+	struct stream *s;
+
+	if (pbrim->install_in_progress)
+		return;
+	zlog_debug("%s: name %s type %d %d", __PRETTY_FUNCTION__,
+		   pbrim->ipset_name, pbrim->type, install);
+	s = zclient->obuf;
+	stream_reset(s);
+
+	zclient_create_header(s,
+			      install ? ZEBRA_IPSET_CREATE :
+			      ZEBRA_IPSET_DESTROY,
+			      VRF_DEFAULT);
+
+	stream_putl(s, 1); /* send one pbr action */
+
+	bgp_encode_pbr_ipset_match(s, pbrim);
+
+	stream_putw_at(s, 0, stream_get_endp(s));
+	if (!zclient_send_message(zclient) && install)
+		pbrim->install_in_progress = true;
+}
+
+void bgp_send_pbr_ipset_entry_match(struct bgp_pbr_match_entry *pbrime,
+				    bool install)
+{
+	struct stream *s;
+
+	if (pbrime->install_in_progress)
+		return;
+	zlog_debug("%s: name %s %d %d", __PRETTY_FUNCTION__,
+		   pbrime->backpointer->ipset_name,
+		   pbrime->unique, install);
+	s = zclient->obuf;
+	stream_reset(s);
+
+	zclient_create_header(s,
+			      install ? ZEBRA_IPSET_ENTRY_ADD :
+			      ZEBRA_IPSET_ENTRY_DELETE,
+			      VRF_DEFAULT);
+
+	stream_putl(s, 1); /* send one pbr action */
+
+	bgp_encode_pbr_ipset_entry_match(s, pbrime);
+
+	stream_putw_at(s, 0, stream_get_endp(s));
+	if (!zclient_send_message(zclient) && install)
+		pbrime->install_in_progress = true;
+}
+
+void bgp_send_pbr_iptable(struct bgp_pbr_action *pba,
+			  struct bgp_pbr_match *pbm,
+			  bool install)
+{
+	struct stream *s;
+
+	if (pbm->install_iptable_in_progress)
+		return;
+	zlog_debug("%s: name %s type %d mark %d %d", __PRETTY_FUNCTION__,
+		   pbm->ipset_name, pbm->type, pba->fwmark, install);
+	s = zclient->obuf;
+	stream_reset(s);
+
+	zclient_create_header(s,
+			      install ? ZEBRA_IPTABLE_ADD :
+			      ZEBRA_IPTABLE_DELETE,
+			      VRF_DEFAULT);
+
+	bgp_encode_pbr_iptable_match(s, pba, pbm);
+
+	stream_putw_at(s, 0, stream_get_endp(s));
+	if (!zclient_send_message(zclient) && install) {
+		pbm->install_iptable_in_progress = true;
+		pba->refcnt++;
+	}
+}
+
+/* inject in table <table_id> a default route to:
+ * - if nexthop IP is present : to this nexthop
+ * - if vrf is different from local : to the matching VRF
+ */
+void bgp_zebra_announce_default(struct bgp *bgp, struct nexthop *nh,
+				afi_t afi, uint32_t table_id, bool announce)
+{
+	struct zapi_nexthop *api_nh;
+	struct zapi_route api;
+	struct prefix p;
+
+	if (!nh || nh->type != NEXTHOP_TYPE_IPV4
+	    || nh->vrf_id == VRF_UNKNOWN)
+		return;
+	memset(&p, 0, sizeof(struct prefix));
+	/* default route */
+	if (afi != AFI_IP)
+		return;
+	p.family = AF_INET;
+	memset(&api, 0, sizeof(api));
+	api.vrf_id = bgp->vrf_id;
+	api.type = ZEBRA_ROUTE_BGP;
+	api.safi = SAFI_UNICAST;
+	api.prefix = p;
+	api.tableid = table_id;
+	api.nexthop_num = 1;
+	SET_FLAG(api.message, ZAPI_MESSAGE_TABLEID);
+
+	/* redirect IP */
+	if (nh->gate.ipv4.s_addr) {
+		char buff[PREFIX_STRLEN];
+
+		SET_FLAG(api.message, ZAPI_MESSAGE_NEXTHOP);
+		api_nh = &api.nexthops[0];
```

Comment:
this and the setting of the nh vrf_id can be moved to outside the if statement.

Reply:
done